documentation/diagnosing-slow-builds.md

```diff
@@ -0,0 +1,44 @@
+# Diagnosing slow builds
+
+No build is ever fast enough. It's often worth investing some effort on understanding where MSBuild spends its time when building your code, and thinking about whether it can be improved.
+
+[Obligatory XKCD](https://www.xkcd.com/303/)
+
+Logs at MSBuild’s “normal” level don't capture most of the really interesting information, though.
+
+Try [capturing a binary log](https://aka.ms/msbuild/binlog) in your build; that preserves detailed information and allows for offline analysis. Changing log verbosity levels can affect build performance, but doesn't usually cause dramatic slowdown for full builds. Cross-check the results of with-logging and without-logging builds to make sure they're not completely out of whack before delving deeply into analyzing the binary log.
+
+## Setting expectations
+
+Do not expect to see 100% processor utilization during the entire build. Builds are often constrained by disk I/O, and the build process is also constrained by the dependencies between projects that are being built. If 100 projects depend on a common library, that library must be built before they can start building.
+
+For this reason, avoid unnecessary dependencies. Dependencies are usually `ProjectReference` items, but can also be expressed in a solution file as "solution build dependencies".
+
+Dependencies between projects also usually keep `-maxcpucount` values from scaling linearly.
+
+## Building in parallel
+
+The biggest performance win is going from single-threaded build to multi-process build by adding `-m` to your command-line build invocation (or equivalent in build automation).
+
+Standard project types should support multiprocessor build without any work on your part, but customizations or very old projects may have race conditions that need to be fixed. It's worth the effort!
```

Comment:
Avoid file references to anything that's an output of your build. It will like cause a race. Use project references OR use manually specified project dependencies in the solution file (advanced -- add these using VS)

Reply:
how do you avoid races? in our build we have dozens of projects in our solution that are dependent on a fairly common project being built first, and I see a bunch of file write errors in the `CopyFilesToOutputDirectory` because other instances of MSBuild have a lock on the file

Reply:
if you add dependencies with VS, does that pretty much mean you can never build from the command line? or do you always have to build the .sln to get consistent results?